### PR TITLE
wpf: add support for VT mouse mode

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
## Summary of the Pull Request

This pull request ports the VT mouse code from TermControl to WpfTerminalControl. Our WPF control is a lot closer to Win32 than to Xaml, so our mouse event handler looks _nothing_ like the one that we got from Xaml. We can pass events through almost directly, because the window message handling in the mouse input code actually came from _conhost_. It's awesome.

Neither TermControl nor conhost pass hover events through when the control isn't focused, so I wired up focus events to make sure we acted the same.

Just like Terminal and conhost, mouse events are suppressed when <kbd>Shift</kbd> is held.

## Validation Steps Performed
Tested with MC, and tested by manually engaging SGR events in an Echo terminal.

image

### CODE REVIEW

**Comment by WSLUser** [comment]
> our mouse event handler looks nothing like the one that we got from Xaml

We can add XAML controls here for it if you really wanted to. https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands#wpf-and-windows-forms-applications

At some point, we probably should start hosting XAML in WPF.

**Comment by DHowett-MSFT** [comment]
@WSLUser yep, it’s a good idea! Right now our main consumer for the WPF control is _Visual Studio_, who can’t use the Islands or WinRR activation machinery UWP XAML requires on all the versions of the OS they run on (like, Windows 7 through 10.0.17763).

File: src/cascadia/PublicTerminalCore/HwndTerminal.cpp [hunk 1]
```diff
@@ -27,6 +35,14 @@ LRESULT CALLBACK HwndTerminal::HwndTerminalWndProc(
 
     if (terminal)
     {
+        if (_IsMouseMessage(uMsg) && terminal->_CanSendVTMouseInput())
```
Comment (ZoeyR) [inline src/cascadia/PublicTerminalCore/HwndTerminal.cpp:38]: I really wish we could use `_IsMouseMessage` in the case statements, I don't like this being separate. But I also wouldn't want to see that large list of message types in the switch 😅

Comment (DHowett-MSFT) [reply src/cascadia/PublicTerminalCore/HwndTerminal.cpp:38]: > Don't like this being separate

I agree totally with all these points. Conhost has them all as case labels, and it's monstrous. I may come through afterwards and refactor this code for flow/order, so we shall get a chance to revisit :smile:

**Pull request merged**